cmd/watch-termination/main.go

```diff
@@ -126,12 +127,31 @@ func run() int {
 			// keep going
 		}
 
-		defer func() {
+		var deleteLockOnce sync.Once
+
+		if *gracefulTerminatioPeriod > 2*time.Second {
+			go func() {
+				<-time.After(*gracefulTerminatioPeriod - 2*time.Second)
+
+				deleteLockOnce.Do(func() {
+					klog.Infof("Graceful termination time nearly passed and kube-apiserver has still not terminated. Deleting termination lock file %q to avoid a false positive.", *terminationLock)
+					if err := os.Remove(*terminationLock); err != nil {
+						klog.Errorf("Termination lock file deletion failed: %v", err)
+					}
+
+					if err := eventf(client.CoreV1().Events(ref.Namespace), *ref, corev1.EventTypeWarning, "GracefulTerminationTimeout", "kube-apiserver did not terminate within %s", gracefulTerminatioPeriod); err != nil {
+						klog.Error(err)
+					}
+				})
+			}()
+		}
+
+		defer deleteLockOnce.Do(func() {
 			klog.Infof("Deleting termination lock file %q", *terminationLock)
```

Comment:
maybe move everything here and do:

```
select {
<-time.After(*gracefulTerminatioPeriod - 2*time.Second):
defalt:
}
```

Reply:
but this does not run when kube-apiserver is not terminated